cpp/rng/rng.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+#include "rng.h"
+#include "../util/error.h"
+#include "../log.h"
+
+#include <chrono>
+#include <random>
+#include <sstream>
+#include <cstdlib>
+#include <string.h>
+
+namespace openage {
+namespace rng {
+
+rng::rng(uint64_t v1) {
+	this->seed(v1);
+}
+
+rng::rng(const void *data, uint32_t len) {
+	this->seed(data, len);
+}
+
+rng::rng(const std::string &instr) {
+	this->from_string(instr);
+}
+
+rng::rng(std::istream &instream) {
+	this->from_stream(instream);
+}
+
+void rng::seed(uint64_t val) {
+	this->seed(&val, sizeof(val));
+}
+
+//murmurhash3
+//seeds must be hashed since they may be poorly distributed
+//and may have high zero content initially
+static uint64_t hash64(uint64_t val) {
+	val ^= val >> 33;
+	val *= 0xff51afd7ed558ccd;
+	val ^= val >> 33;
+	val *= 0xc4ceb9fe1a85ec53;
+	return val ^ (val >> 33);
+}
+
+static void hash_seed_from_blob(const void *dat, uint64_t *state, int64_t bytes) {
+	const char *data = static_cast<const char *>(dat);
+	uint64_t uint64 = 0;
+	memcpy(&uint64, data, bytes <= 8 ? bytes : 8);
+	state[0] = hash64(uint64);
+	state[1] = hash64(uint64 ^ state[0]);
+	for(int64_t i = 1; i <= bytes - 8; i++) {
+		state[0] = hash64(state[0] ^ *reinterpret_cast<const uint64_t *>(data + i));
+		state[1] = hash64(state[0] ^ state[1]);
+	}
+	
+	if(!(state[0] || state[1])) {
+		state[0] = 0x0123456789abcdef;
+	}
+}
+
+void rng::seed(const void *data, uint32_t bytes) {
+	if(bytes == 0) {
+		throw util::Error("0 bytes supplied as seed data");
+	}
+	hash_seed_from_blob(data, this->state, bytes);
+	this->discard(discard_on_seed);
+}
+
+//http://en.wikipedia.org/wiki/Xorshift
+//!actually does the number crunching
+static inline uint64_t do_rng(uint64_t &s0, uint64_t &s1) {
```

Comment:
Nice, inlined. Is it intended that `s0` and `s1` get modified? If not, can `&s0` and `&s1` be `const &`?